nemo/collections/nlp/models/question_answering/qa_bert_model.py

```diff
@@ -0,0 +1,892 @@
+        # and `pred_text`, and check if they are the same length. If they are
+        # NOT the same length, the heuristic has failed. If they are the same
+        # length, we assume the characters are one-to-one aligned.
+        tokenizer = BasicTokenizer(do_lower_case=do_lower_case)
+
+        tok_text = " ".join(tokenizer.tokenize(orig_text))
+
+        start_position = tok_text.find(pred_text)
+        if start_position == -1:
+            if verbose_logging:
+                logging.warning("Unable to find text: '%s' in '%s'" % (pred_text, orig_text))
+            return orig_text
+        end_position = start_position + len(pred_text) - 1
+
+        (orig_ns_text, orig_ns_to_s_map) = _strip_spaces(orig_text)
+        (tok_ns_text, tok_ns_to_s_map) = _strip_spaces(tok_text)
+
+        if len(orig_ns_text) != len(tok_ns_text):
+            if verbose_logging:
+                logging.warning(
+                    "Length not equal after stripping spaces: '%s' vs '%s'", orig_ns_text, tok_ns_text,
+                )
+            return orig_text
+
+        # We then project the characters in `pred_text` back to `orig_text` using
+        # the character-to-character alignment.
+        tok_s_to_ns_map = {}
+        for (i, tok_index) in tok_ns_to_s_map.items():
+            tok_s_to_ns_map[tok_index] = i
+
+        orig_start_position = None
+        if start_position in tok_s_to_ns_map:
+            ns_start_position = tok_s_to_ns_map[start_position]
+            if ns_start_position in orig_ns_to_s_map:
+                orig_start_position = orig_ns_to_s_map[ns_start_position]
+
+        if orig_start_position is None:
+            if verbose_logging:
+                logging.warning("Couldn't map start position")
+            return orig_text
+
+        orig_end_position = None
+        if end_position in tok_s_to_ns_map:
+            ns_end_position = tok_s_to_ns_map[end_position]
+            if ns_end_position in orig_ns_to_s_map:
+                orig_end_position = orig_ns_to_s_map[ns_end_position]
+
+        if orig_end_position is None:
+            if verbose_logging:
+                logging.warning("Couldn't map end position")
+            return orig_text
+
+        output_text = orig_text[orig_start_position : (orig_end_position + 1)]
+
+        return output_text
+
+    def _get_raw_scores(self, examples: List, preds: Dict[str, str]):
+        """
+        Computes the exact and f1 scores from the examples and the model predictions
+        """
+        exact_scores = {}
+        f1_scores = {}
+
+        for example in examples:
+            qas_id = example.qas_id
+            gold_answers = [answer["text"] for answer in example.answers if QAMetrics.normalize_answer(answer["text"])]
+
+            if not gold_answers:
+                # For unanswerable questions, only correct answer is empty string
+                gold_answers = [""]
+
+            if qas_id not in preds:
```

Comment:
I wonder when this will be triggered because this looks like a bug if we ever get here